### PR TITLE
[SUPERCEDED] Standardize and disambiguate the return from `digest.Parse`

### PR DESCRIPTION
**Pull request by nima**
~~Please see https://github.com/oras-project/oras-go/issues/225 and https://github.com/oras-project/oras-go/pull/237 for background.~~

Superseded by #77

### CODE REVIEW

File: verifiers.go [hunk 1]
```diff
@@ -42,5 +42,10 @@ func (hv hashVerifier) Write(p []byte) (n int, err error) {
 }
 
 func (hv hashVerifier) Verified() bool {
-	return hv.digest == NewDigest(hv.digest.Algorithm(), hv.hash)
+	switch digest := NewDigest(hv.digest.Algorithm(), hv.hash); digest {
+	case hv.digest: return true
+	case FromString(""): return true
+	}
+
+	return false
```
Comment (thaJeztah) [inline verifiers.go:50]: I think it's more common to put the return's on a new line; perhaps also include the last return inside the switch itself;

```go
switch digest := NewDigest(hv.digest.Algorithm(), hv.hash); digest {
case hv.digest:
	return true
case FromString(""):
	return true
default:
	return false
}
```

**Review by thaJeztah — CHANGES_REQUESTED** [review]
I left some comments, but I'm a bit confused about the change itself; IIUC from the linked issue, the issue you're trying to resolve is a registry that doesn't return the `Docker-Content-Digest` header. The code in this repository only provides the tools to perform the validation, but does not handle reading the `Docker-Content-Digest` specifically (i.e., greping for that string does not produce a result in the source code).

Note that the header is _optional_ but _if_ it's returned by the registry, it should contain a valid digest. From the spec;

https://github.com/opencontainers/distribution-spec/blob/4ab4752c3b86a926d7e5da84de64cbbdcc18d313/spec.md#legacy-docker-support-http-headers

> ### Legacy Docker support HTTP headers
>
> Because of the origins this specification, the client MAY encounter Docker-specific headers, such as `Docker-Content-Digest`, or `Docker-Distribution-API-Version`. These headers are OPTIONAL and clients SHOULD NOT depend on them.

_However_, IF the header is present, and IF the client uses the header, it MUST be validated;

https://github.com/opencontainers/distribution-spec/blob/4ab4752c3b86a926d7e5da84de64cbbdcc18d313/spec.md#pulling-manifests

> A GET request to an existing manifest URL MUST provide the expected manifest, with a response code that MUST be `200 OK`.
> A successful response SHOULD contain the digest of the uploaded blob in the header `Docker-Content-Digest`.
> 
> The `Docker-Content-Digest` header, if present on the response, returns the canonical digest of the uploaded blob which MAY differ from the provided digest.
> If the digest does differ, it MAY be the case that the hashing algorithms used do not match.
> See [Content Digests](https://github.com/opencontainers/image-spec/blob/v1.0.1/descriptor.md#digests) <sup>[apdx-3](#appendix)</sup> for information on how to detect the hashing algorithm in use.
> Most clients MAY ignore the value, but if it is used, the client MUST verify the value against the uploaded blob data.


So if the registry is not returning a `Docker-Content-Digest`, no validation is performed with it, but if it _does_ return the header, it should be valid; which means that if the registry returns an empty value, there's a bug in that registry implementation, and it doesn't comply with the specification.

File: digest.go [hunk 2]
```diff
@@ -80,7 +80,10 @@ var (
 // be returned if the format is invalid.
 func Parse(s string) (Digest, error) {
 	d := Digest(s)
-	return d, d.Validate()
+	if err := d.Validate(); err != nil {
+		return "", err
+	}
```
[the remark this answers is not in the document]
Comment (nima) [reply digest.go:85]: Good catch @thaJeztah thank you.  While I was chasing that down, I discovered more issues with the approach taken in `oras-go` (I've started a discussion for that separately [here](https://cloud-native.slack.com/archives/CJ1KHJM5Z/p1657935407555609)).  Once I have a clear understanding, I'll revisit this PR and send an update.

Thanks for taking a look!

File: digest.go [hunk 3]
```diff
@@ -80,7 +80,10 @@ var (
...
+    return d, nil
```
[the remark this answers is not in the document]
Comment (nima) [reply digest.go:86]: Hmmm, I thought GoLand takes care of all that, I'll investigate.

File: verifiers.go (hunk 1 above)
Comment (nima) [reply verifiers.go:50]: Noted! :)

**Comment by nima** [comment]
Thank you again for the detailed review here @thaJeztah.

I agree with everything you've pointed out here, I'll revisit my changes in this repository and then either a) revert the change, or b) get back to you with why I think the change should go it.

> I left some comments, but I'm a bit confused about the change itself; IIUC from the linked issue, the issue you're trying to resolve is a registry that doesn't return the `Docker-Content-Digest` header. The code in this repository only provides the tools to perform the validation, but does not handle reading the `Docker-Content-Digest` specifically (i.e., greping for that string does not produce a result in the source code).
> 
> Note that the header is _optional_ but _if_ it's returned by the registry, it should contain a valid digest. From the spec;
> 
> https://github.com/opencontainers/distribution-spec/blob/4ab4752c3b86a926d7e5da84de64cbbdcc18d313/spec.md#legacy-docker-support-http-headers
> 
> > ### Legacy Docker support HTTP headers
> > Because of the origins this specification, the client MAY encounter Docker-specific headers, such as `Docker-Content-Digest`, or `Docker-Distribution-API-Version`. These headers are OPTIONAL and clients SHOULD NOT depend on them.
> 
> _However_, IF the header is present, and IF the client uses the header, it MUST be validated;
> 
> https://github.com/opencontainers/distribution-spec/blob/4ab4752c3b86a926d7e5da84de64cbbdcc18d313/spec.md#pulling-manifests
> 
> > A GET request to an existing manifest URL MUST provide the expected manifest, with a response code that MUST be `200 OK`.
> > A successful response SHOULD contain the digest of the uploaded blob in the header `Docker-Content-Digest`.
> > The `Docker-Content-Digest` header, if present on the response, returns the canonical digest of the uploaded blob which MAY differ from the provided digest.
> > If the digest does differ, it MAY be the case that the hashing algorithms used do not match.
> > See [Content Digests](https://github.com/opencontainers/image-spec/blob/v1.0.1/descriptor.md#digests) [apdx-3](#appendix) for information on how to detect the hashing algorithm in use.
> > Most clients MAY ignore the value, but if it is used, the client MUST verify the value against the uploaded blob data.
> 
> So if the registry is not returning a `Docker-Content-Digest`, no validation is performed with it, but if it _does_ return the header, it should be valid; which means that if the registry returns an empty value, there's a bug in that registry implementation, and it doesn't comply with the specification.

File: digest.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (nima) [reply digest.go]: Not sure how that crept in, but no `/^ /` anymore.

**Comment by nima** [comment]
Hi @thaJeztah, when you get some time could you please take a look at this change and give me your feedback and list of concerns?

**Comment by AkihiroSuda** [comment]
The PR title and the description text is confusing.
Hard to understand what this PR is.

File: digest.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (AkihiroSuda) [reply digest.go:85]: Why not keep the previous code?

File: digest.go [hunk 4]
```diff
@@ -137,11 +140,21 @@ func (d Digest) Encoded() string {
 	return string(d[d.sepIndex()+1:])
 }
 
-// Hex is deprecated. Please use Digest.Encoded.
+// IsIdenticalTo compares this digest to any other, inclusive of the hashing algorithm used
+func (d Digest) IsIdenticalTo(o Digest) bool {
+	return string(d) == string(o)
+}
```
Comment (AkihiroSuda) [inline digest.go:146]: This should be probably a separate PR

File: digest.go [hunk 5]
```diff
@@ -137,11 +140,21 @@ func (d Digest) Encoded() string {
...
+
+// Deprecated: Hex has been deprecated in favor of Digest.Encoded
 func (d Digest) Hex() string {
 	return d.Encoded()
 }
 
+// IsEmpty does what it says (an empty digest is an empty string)
+func (d Digest) IsEmpty() bool {
```
Comment (AkihiroSuda) [inline digest.go:154]: This one too

**Comment by nima** [comment]
I've dropped the dependency on this change in `oras-go`, but I wonder if it still makes sense to keep this change.  a) returning a payload + an error seems like something that could lead to confusion, but probably generally safe (since callers generally first check err).  b) Do you think the helpers are adding value, or extra code that's not really worth the added LoC?

**Comment by nima** [comment]
> The PR title and the description text is confusing. Hard to understand what this PR is.

I'll clean this up now and split it into two separate PRs.

**Comment by nima** [comment]
This was my first PR, and it was (as pointed out) confusing, poorly documented, and contained more than one target change.

Superseded by #77
